src/polyit_rage/utils/exceptions.py

```
@@ -0,0 +1,120 @@
+"""
+Custom exceptions for the PolyitRAGE system.
+
+Provides specific exception types for different error scenarios
+to enable precise error handling and logging.
+"""
+
+
+class PolyitRAGEError(Exception):
+    """Base exception for all PolyitRAGE errors."""
+    
+    def __init__(self, message: str, details: dict = None):
```

Ruff (0.14.11)

PEP 484 prohibits implicit `Optional`

Convert to `T | None`

(RUF013)